x/ibc/04-channel/keeper/handshake.go

```diff
@@ -33,38 +35,45 @@ func (k Keeper) ChanOpenInit(
 	connectionHops []string,
 	portID,
 	channelID string,
+	portCap capability.Capability,
 	counterparty types.Counterparty,
 	version string,
-) error {
+) (capability.Capability, error) {
 	// channel identifier and connection hop length checked on msg.ValidateBasic()
 
 	_, found := k.GetChannel(ctx, portID, channelID)
 	if found {
-		return sdkerrors.Wrap(types.ErrChannelExists, channelID)
+		return nil, sdkerrors.Wrap(types.ErrChannelExists, channelID)
 	}
 
 	connectionEnd, found := k.connectionKeeper.GetConnection(ctx, connectionHops[0])
 	if !found {
-		return sdkerrors.Wrap(connection.ErrConnectionNotFound, connectionHops[0])
+		return nil, sdkerrors.Wrap(connection.ErrConnectionNotFound, connectionHops[0])
 	}
 
 	if connectionEnd.GetState() == connectionexported.UNINITIALIZED {
-		return sdkerrors.Wrap(
+		return nil, sdkerrors.Wrap(
 			connection.ErrInvalidConnectionState,
 			"connection state cannot be UNINITIALIZED",
 		)
 	}
 
+	if !k.scopedKeeper.AuthenticateCapability(ctx, portCap, portID) {
+		return nil, sdkerrors.Wrap(porttypes.ErrInvalidPort, "caller does not own port capability")
+	}
+
 	channel := types.NewChannel(exported.INIT, order, counterparty, connectionHops, version)
 	k.SetChannel(ctx, portID, channelID, channel)
 
-	// TODO: blocked by #5542
-	// key := ""
-	// k.SetChannelCapability(ctx, portID, channelID, key)
+	capKey, err := k.scopedKeeper.NewCapability(ctx, channelID)
```

Comment:
This is definitely wrong but it seemed like what the TODO and spec wanted me to do?

Can't create a dynamic capability key and then persist into state since that will cause nodes to disagree on apphash.
Yet spec asks the following to happen here:

```
key = generate()
provableStore.set(channelCapabilityPath(portIdentifier, channelIdentifier), key)
```

Seems like either the generate() function needs to provide a static capability key or we cannot set it in the provable store (which from my understanding is the CommitMultiStore in the SDK)

Should I be generating a dynamic key and just leaving it in the memorystore? Or create a static key here?

Reply:
See updated spec [here](https://github.com/cosmos/ics/tree/cwgoes/capability-rework/spec/ics-004-channel-and-packet-semantics#opening-handshake):

```typescript
    key = newCapability(channelCapabilityPath(portIdentifier, channelIdentifier))
    return key
```

`SetChannelCapability` can be completely removed, it is no longer needed.